keps/sig-scheduling/3280-guarantee-pdb-when-preemption-happens/README.md

```diff
@@ -0,0 +1,970 @@
+which indicates that the scheduler will try to avoid violating the PDB during preemption, 
+but it cannot be guaranteed. The other is `PodDisruptionBudgetStrict` which indicates that the 
+PDB will not be violated during scheduler preemption, as noted above, if the preempter 
+has a priority ClassName of `system-cluster-critical` or `system-node-critical`, it may 
+still potentially violate the victim's PDB. New policies may be added in the future based 
+on the need, but it will not be discussed in this article at this time.
+
+```go
+// DisruptionPolicy specifies how pods with this priority class 
+// may be disrupted when being preempted.
+type DisruptionPolicy string
+
+const (
+	// PodDisruptionBudgetStrict indicates when being preempted by a higher priority pod, 
+	// strict adherence to the PodDisruptionBudget of the pod with this priority class is required.
+	// But it the preempter with these priority classes of `system-cluster-critical` or `system-node-critical` 
+	// will still be able to violate this pod's PDB during preemption.
+	PodDisruptionBudgetStrict PreemptedPolicy = "PodDisruptionBudgetStrict"
+	
+	// PodDisruptionBudgetBestEffort indicates when being preempted by a higher priority pod, 
+	// an effort is made to adhere to the PodDisruptionBudget of the pod with this priority class
+	// as much as possible.
+	PodDisruptionBudgetBestEffort PreemptedPolicy = "PodDisruptionBudgetBestEffort"
+)
+```
+
+The following is an example of a `PriorityClass` (PDB) where the user sets the `disruptionPolicy` 
+to `PodDisruptionBudgetStrict` indicating that the PDB for the zookeeper must not be violated during the 
+scheduler preemption process.
+
+```yaml
+apiVersion: scheduling.k8s.io/v1
+kind: PriorityClass
+metadata:
+  name: low-priority
+value: 100
+globalDefault: false
+description: "This priority class should be used for XYZ service pods only."
+disruptionPolicy: PodDisruptionBudgetStrict
+```
+
+The scheduler plugin `defaultpreemption` needs to check the policy set in the "preemptionPolicy" field 
+when selecting victims.
+- if it's set to `PodDisruptionBudgetBestEffort`, the implementation will remain consistent with the existing behavior,
+  meaning that the scheduler will try to select victims whose PDBs are not violated by preemption, but 
+  if no such victims are found, preemption will still happen and lower priority pods will be removed 
+  despite their PDBs being violated.
+- if it's set to `PodDisruptionBudgetStrict`, there are two options:
+  - If the preempter's priority ClassName is `system-cluster-critical` or `system-node-critical` the implementation 
+    will be the same as `PodDisruptionBudgetBestEffort`.
+  - If not, the scheduler will check if the victim' PDBs will be violated when selecting victims
+    - if violate the victims' PDBs, this victim will not be selected as candidates.
+    - if not violate the victims' PDBs, scheduler will preempt this pod via the `/evictions` endpoint.
```

Comment:
Please expand on what happens if the scheduler sends a call through `/evictions` but the pod is not deleted. Would the incoming pod be able to choose to preempt a different pod in the next iteration?

What would cause the scheduler to retry this pod if `/evictions` rejects the deletion? Since there is no pod being deleted, is there any event the scheduler can react to?

Reply:
one of the ways would be react to the HTTP code of the eviction request. 429 (`TooManyRequests`) comes to mind but it also gets returned when `DisruptionsAllowed == 0` so we would either have to distinguish that via the message, or locally somehow, or distinguish & update the error responses in the eviction API

Reply:
If it responds `200 OK`,  it means the eviction is allowed, and the victim is deleted, similar to sending a DELETE request to the Pod URL. If it responses  `429 Too Many Requests` or `500 Internal Server Error`, the scheduler will output an error log and terminate the preemption process (the same as the current behavior of deletion failure).

Reply:
> Would the incoming pod be able to choose to preempt a different pod in the next iteration?

Maybe don't need，In addition to the apiserver service network issue, failure may only occur in the event of concurrent issues between the scheduler and controller.

Reply:
> and terminate the preemption process

My question is about what happens after that. The pod should be entering the unschedulable queue. What would bring it back into the active queue? Or would it stay in the active/backoff queue?

Reply:
FYI, I meant in these two cases it is possible to retry eviction later.

- https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L301
- https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L390

not sure if it is necessary to mention 500 error code in the enhancement as there are other possible error codes as well

Reply:
> My question is about what happens after that. The pod should be entering the unschedulable queue. What would bring it back into the active queue? Or would it stay in the active/backoff queue?

get it. 

You are right, if `/evictions` interface is called during the preemption process, but the responses is `429 Too Many Requests`. Assuming no additional events arrive, the preemptor cannot be moved to the activeQ until it is flushed. 

So I think we should not terminate the preemption process directly, but choose another victim among the candidate victims(remove the previous victims from victimsMap and run pickOneNodeForPreemption again) to preempt until it succeeds or there are no more candidates.